### DISCUSSION TITLE
Add translations for texts from campaign scenario's maps

### DISCUSSION THREAD

**Discussion by Effektus** [ihhub/fheroes2#6872] [Ideas]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

Missing translation string in start of campaign

### Save file

![image](https://user-images.githubusercontent.com/12934196/201635457-8dd62503-2e36-4996-bab2-2dcabefd5f63.png)

**Comment by zenseii** [comment]
Hi @Effektus.

These dialogs that pop up while in a scenario whether campaign or not, are baked into the file itself and not part of the engine. We cannot modify and distribute the map files due to copyright and licensing issues.

As of now there is no way to translate these with the engine.

**Comment by Effektus** [comment]
Ок

**Comment by undef21** [comment]
@zenseii Could we use external translations for maps? For example, if we have the file CAMP1_01.HXC, then we  place CAMP1_01.HXC.bg, CAMP1_01.HXC.ru ... translation files beside.

**Comment by zenseii** [comment]
Hi @undef21.
I think @ihhub would be more qualified to answer this. If we can come up with a solution that allows this and is in the clear then I'm fine with it.

Do you propose to have a separate file that overwrites the content of the map file on runtime?

If we add files for every campaign mission and scenario then these would become too many. Not to mention the inconvenience of working with several files with translation compared to a single one like currently

**Comment by ihhub** [comment]
First and foremost, a citation from licensing file provided by Ubisoft:
```
It is not permitted:
...
-    To create or distribute unauthorised levels and/or scenarios,
...
```
Proving text translation for maps comes with an indirect violation of the statement item because the text is a part of the map.

Secondly, what will happen if a user has maps from German, Italian, Polish or Russian version of the game but they want to play in their own language? How are we going to handle this? Asking user to buy another copy of the game is out of the question. Providing a half-baked solution is a very bad approach of software development.

Thirdly, people will ask why campaigns are translated but individual maps aren't.

**Comment by undef21** [comment]
> Do you propose to have a separate file that overwrites the content of the map file on runtime?

Yes. But ihhub has already explained licensing pitfalls.

**Comment by ihhub** [comment]
I am moving this into discussion section as we cannot do anything for such an issue. The future campaign for the upcoming add-on will properly support all translations.

**Comment by Mr-Bajs** [comment]
But what's the plan for the new editor and the new mapformat regaring translations?
Will the new mapformat support translations? It would be neat if it did.

**Comment by dood-apo** [comment]
@ihhub how about something like this?
```
diff --git a/src/fheroes2/game/game_startgame.cpp b/src/fheroes2/game/game_startgame.cpp
index 47546b74..613e323e 100644
--- a/src/fheroes2/game/game_startgame.cpp
+++ b/src/fheroes2/game/game_startgame.cpp
@@ -966,7 +966,7 @@ fheroes2::GameMode Interface::AdventureMap::HumanTurn( const bool isload )
                                                fheroes2::Text( event.message, fheroes2::FontType::normalWhite() ), Dialog::OK, funds );
             }
             else if ( !event.message.empty() ) {
-                fheroes2::showStandardTextMessage( event.title, event.message, Dialog::OK );
+                fheroes2::showStandardTextMessage( event.title, _(event.message), Dialog::OK );
             }
         } );
```
This would open up a possibility of translating messages present in the map files without licensing issues (at least that is my understanding). With such an approach we could also provide a small resource bundle with erratas for spelling/grammar mistakes in English from the original maps.

Example (possible content of `en.po` fixing a mistake in "Who am I?" standard scenario):
msgid "Scouts have reported a town to the North.  They have also heard some disturbing news.  **Their** seems to be a war raging in this area."
msgstr "Scouts have reported a town to the North.  They have also heard some disturbing news.  **There** seems to be a war raging in this area."

**Reply by oleg-derevenetz** [reply]
Technically this is possible, but there is one potential licensing issue - we will still need the original texts from these maps baked in corresponding .po files. Also we will need to generate these .po files, but we can't do this in our CI for obvious reasons, so we will need to ask someone to generate them separately.

**Reply by dood-apo** [reply]
Got it. I am not really into licensing stuff, so I'll take your word for it that quoting messages from the dialogs in the `*.po` files is risky.

**Reply by oleg-derevenetz** [reply]
@dood-apo I don't know is it risky, it's up to @ihhub to decide. But what is 100% risky - it's generation of .po files from campaign maps in our CI :)))

**Reply by dood-apo** [reply]
Of course. Never suggested it. I just wasn't aware of the fact that adding manual entries (msgid+msgstr pairs) to the `*.po` files is not accepted. I guess the reason is it would be too hard to review and test the translations if we wouldn't be 100% sure that the msgids are correct in the translation PRs.

**Comment by Mr-Bajs** [comment]
Translated maps and campains would be nice. However translated campaigns and maps that aren't the original or created with the OG map editor, should pose no problem. But since the editor isn't done and there are no maps och campaigns created with the new editor at the moment, i suppose that something for a future disscussion.

**Comment by altiereslima** [comment]
Wouldn't it be possible to use the VCMI approach by using mods?

**Comment by AlexSnowLeo** [comment]
I can only support this initiative. Map modifications, this is not really a modification of the game engine. In my understanding, this is just a newly made map (company) with a translate included in it. If the map editor supports the creation of companies, then I assure you, someone will definitely adapt the current companies and translate them into other languages.

As for the translation, for example, into Russian.
There is a version of HMM2 released by Buka. And there, the company maps are translated into Russian, among other things. You can use these resources with the current fheroes2 engine as well. Unless, of course, you have a game license from the Buka company.
![image](https://github.com/user-attachments/assets/c485ee7d-07ca-4507-bd49-d2918db3f97d)

The same can be said about the resources of maps officially released by the 3DO company - German, Polish and French. You do not need to buy these licenses separately, they are all officially present in the general license by GOG.com for example. You need only ".\MAPS\CAMP***.h2c" files.

**Comment by felix-schneider** [comment]
An alternative approach would be to move away from having the translation key be the original english text, but an identifier (c.f. Factorio: https://wiki.factorio.com/Tutorial:Localisation). This would be a big change, requiring all strings in the code to be replaced, but it would eliminate the need to ship the original strings in an English localization file.

There would still need to be a mapping between the English string and the identifier, maybe through a hash table?
Sounds like a lot of work for not a huge benefit, but if there is no other way to have localization without copyright issues, that may be the only way.

Of course, when we are talking about localizing map strings, do we mean taking the actual strings from the other language versions of the OG (for those languages that have their own version)? Because that would equally be a copyright issue.

**Reply by zenseii** [reply]
Hi, @felix-schneider. As pointed out in Discord the issue regarding original texts is much smaller than it might seem. fheroes2 does not contain 100% all the original strings from the original English Heroes 2, because a lot of them have either been completely replaced with better strings, or changed in some way to create consistency in the game, like using the same expressions for things, adding explanations about hidden mechanics etc.. These changes will naturally also be reflected in the various translations. This is only the current state, many many more of the original strings will be changed or replaced entirely into something much better. Indeed we encourage translators to improve upon and modernize the original localization, because many things have changed since 1996.

As an example, the French translation provided with fheroes2 is very different from the original one. It aims to use the same terminology as the translations used in the later games, making the experience much more consistent.

Regardless of this, fheroes2 aims to add accessibility features to the game, making it possible for people with disabilities to enjoy the game as much as people without them. In the EU, which is where Ubisoft's HQs are, there is an exception to copyright in this case for non-profit bodies (the fheroes2 team doesn't even accept donations at this moment, we do not get paid): https://www.copyrightuser.org/understand/exceptions/disability/. We have been reached out to by people with visual impairments about making this game accessible for them. So, to make the game accessible, we had to use text from the original game UI.

We have many examples of work we have done which make fheroes2 accessible, one WIP task is this one https://github.com/ihhub/fheroes2/issues/5270 (_not up to date_), but we have also made work which will make it possible to have Closed Captions (CC) available in the future, not to mention expanding the hotkeys available so that people who need special input devices can enjoy the game.

I hope this clears up some doubts about this aspect. I should also point out that a computer software like Heroes 2 is composed of many copyrighted parts, like music, videos, graphical work, and not just text strings. We don't provide anything else other than some of the original UI text, which is the bare minimum necessary to recreate the game and it is ultimately a very small part of the entire Heroes 2 copyrighted work. As a final note, we are not lawyers and none of this should be taken as legal advice.

**Comment by altiereslima** [comment]
How about, instead of having the map editor save the translated text directly in the map, the editor saves translations in a separate configuration file without editing the map?

**Comment by ihhub** [comment]
The fheroes2 Editor will support multiple languages per map. I suggest to leave the original maps as they are due to copyright issues and wait till our Editor will have the required functionality.
